File: lib/fluxsource/kryofluxfluxsource.cc
#include "lib/core/globals.h"
#include "lib/data/fluxmap.h"
#include "lib/data/locations.h"
#include "lib/external/kryoflux.h"
#include "lib/fluxsource/fluxsource.pb.h"
#include "lib/fluxsource/fluxsource.h"
#include <filesystem>

class KryofluxFluxSource : public TrivialFluxSource
{
public:
    KryofluxFluxSource(const KryofluxFluxSourceProto& config):
        _path(config.directory())
    {
        std::vector<CylinderHead> chs;
        for (const auto& di : std::filesystem::directory_iterator(_path))
        {
            static const std::regex FILENAME_REGEX("([0-9]+)\\.([0-9]+)\\.raw");

            std::string filename = di.path().filename();
            std::smatch dmatch;
            if (std::regex_match(filename, dmatch, FILENAME_REGEX))
                chs.push_back(CylinderHead{(unsigned)std::stoi(dmatch[1]),
                    (unsigned)std::stoi(dmatch[2])});
        }
        _extraConfig.mutable_drive()->set_tracks(
            convertCylinderHeadsToString(chs));
    }

public:
    std::unique_ptr<const Fluxmap> readSingleFlux(int track, int side) override
    {
        return readStream(_path, track, side);
    }

    void recalibrate() override {}

private:
    const std::string _path;
};

std::unique_ptr<FluxSource> FluxSource::createKryofluxFluxSource(
    const KryofluxFluxSourceProto& config)
{
    return std::make_unique<KryofluxFluxSource>(config);
}

File: java/com/cowlark/fluxengine/fluxsource/KryofluxFluxSource.java
package com.cowlark.fluxengine.fluxsource;

import com.cowlark.fluxengine.config.ConfigBuilder;
import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.data.CylinderHead;
import com.cowlark.fluxengine.data.Fluxmap;
import com.cowlark.fluxengine.data.Locations;
import com.cowlark.fluxengine.external.Kryoflux;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * A flux source which reads raw Kryoflux stream files, ported from
 * lib/fluxsource/kryofluxfluxsource.cc.
 */
public class KryofluxFluxSource extends TrivialFluxSource
{
    private static final Pattern FILENAME_REGEX =
            Pattern.compile(".*[^0-9]([0-9]+)\\.([0-9]+)\\.raw");

    private final String path;
    protected ConfigProto extraConfig;

    public KryofluxFluxSource(KryofluxFluxSourceProto config)
    {
        path = config.getDirectory();

        List<CylinderHead> chs = new ArrayList<>();
        File[] files = new File(path).listFiles();
        if (files != null)
        {
            for (File f : files)
            {
                Matcher m = FILENAME_REGEX.matcher(f.getName());
                if (m.matches())
                    chs.add(new CylinderHead(
                            Integer.parseInt(m.group(1)),
                            Integer.parseInt(m.group(2))));
            }
        }

        ConfigProto.Builder builder = ConfigProto.newBuilder();
        builder.getDriveBuilder().setTracks(Locations.convertCylinderHeadsToString(chs));
        extraConfig = builder.build();
    }

    @Override
    public void adjustConfig(ConfigBuilder configBuilder)
    {
        configBuilder.mergeConfig(extraConfig);
    }

    @Override
    public Fluxmap readSingleFlux(FluxReadParameters parameters)
    {
        return Kryoflux.readStream(path, parameters.cylinder(), parameters.head());
    }
}